proposals/0000-borrow-operator.md

```diff
@@ -0,0 +1,195 @@
+# Borrow Operator, Pass-by-Immutable-Borrow, and Pass-by-Value
+
+* Proposal: [SE-NNNN](NNNN-filename.md)
+* Authors: ...
+* Review Manager: TBD
+* Status: ...
+
+<!--
+
+*During the review process, add the following fields as needed:*
+
+* Implementation: [apple/swift#NNNNN](https://github.com/apple/swift/pull/NNNNN) or [apple/swift-evolution-staging#NNNNN](https://github.com/apple/swift-evolution-staging/pull/NNNNN)
+* Decision Notes: [Rationale](https://forums.swift.org/), [Additional Commentary](https://forums.swift.org/)
+* Bugs: [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN), [SR-MMMM](https://bugs.swift.org/browse/SR-MMMM)
+* Previous Revision: [1](https://github.com/apple/swift-evolution/blob/...commit-ID.../proposals/NNNN-filename.md)
+* Previous Proposal: [SE-XXXX](XXXX-filename.md)
+
+-->
+
+## Introduction
+
+In this document, we propose adding a new operator to the language, marked by
+the context sensitive keyword `borrow`, that causes the compiler to pass vars
+and inout arguments to callees as arguments using a `pass-by-immutable-borrow`
+convention instead of a `pass-by-value` convention.
+
+Swift-evolution thread: [Discussion thread topic for that proposal](https://forums.swift.org/)
+
+## Motivation
+
+Swift normally passes vars as non-`inout` parameters to callees using a
+`pass-by-value` convention. A `pass-by-value` argument convention is defined
+formally by the `var` being implicitly copied at the relevant callsite so that
+each argument appears to the callee as formally independent values. As an
+example, consider the following code:
+
+```
+func useValueTwice(_ x: Type, _ y: inout Type) {}
+
+var a: Type = ...
+useValueTwice(a, &a)
+```
+
+Formally during evaluation of `useValueTwice`, the compiler will perform an
+immutable exclusive access to `a`'s storage and copy `a` before beginning a
+mutable exclusive access to `a`'s storage. The reason that function evaluation
+is performed in this way is that otherwise, we could suffer from several
+undesireable outcomes, for example:
+
+1. Potentially trigger exclusivity violations. For instance, if we formally
+accessed the inout parameters memory before performing the `pass-by-value`, we
+would have an exclusivity violation when attempting to exclusively access the
+`var` memory to perform our copy.
+
+2. If we did not copy the parameter, we would need to either perform a `take` of
+the non-`inout` parameter meaning that said memory would now be
+invalidated. This would mean that we couldn't pass in var storage directly as an
+`inout` parameter since the memory would be invalidated.
```

Comment:
The above is awkward.  `either` is missing an `or`.   Rather than "perform a `take`" (an undefined term at this point), I would more conversationally explain that passing a reference for the first argument would result in two references within the function to the same storage and that this "aliasing" is a well-known problem in programming languages such as C that permit unchecked references to be passed to functions.